chips/lowrisc/src/spi_host.rs

```diff
@@ -243,45 +248,78 @@ impl SpiHost {
             } else {
                 self.rx_buf.replace(rx_buf);
                 //Theres more to transfer, continue writing from the offset
-                self.spi_transfer_progress();
+                if let Err(err) = self.spi_transfer_progress() {
+                    //Transfer failed, lets clean up
+                    //Clear all pending errors.
+                    self.clear_err_interrupt();
+                    //Something went wrong, reset IP and clear buffers
+                    self.reset_spi_ip();
+                    self.reset_internal_state();
+                    self.client.map(|client| match self.tx_buf.take() {
+                        None => (),
+                        Some(tx_buf) => client.read_write_done(
+                            tx_buf,
+                            self.rx_buf.take(),
+                            self.tx_offset.get(),
+                            Err(err),
+                        ),
+                    });
```

Comment:
we shouldn't call an upcall from within a downcall like this

Reply:
> we shouldn't call an upcall from within a downcall like this

Ah okay, I'll fix this up! What is the reasoning for this?

Reply:
It is justified here: https://github.com/tock/tock/blob/master/doc/reference/trd2-hil-design.md#rule-1-dont-make-synchronous-callbacks

Reply:
thanks!